wallet-new/src/Cardano/Wallet/Kernel/DB/TxMeta/Types.hs

```diff
@@ -0,0 +1,173 @@
+        , t1 ^. txMetaIsOutgoing == t2 ^. txMetaIsOutgoing
+        ]
+
+-- | Lenient equality for two 'TxMeta': two 'TxMeta' are equal if they have
+-- the same data, even if in different order.
+-- NOTE: This check might be slightly expensive as it's logaritmic in the
+-- number of inputs & outputs, as it requires sorting.
+isomorphicTo :: TxMeta -> TxMeta -> Bool
+isomorphicTo t1 t2 =
+    and [ t1 ^. txMetaId == t2 ^. txMetaId
+        , t1 ^. txMetaAmount == t2 ^. txMetaAmount
+        , NonEmpty.sort (t1 ^. txMetaInputs)  == NonEmpty.sort (t2 ^. txMetaInputs)
+        , NonEmpty.sort (t1 ^. txMetaOutputs) == NonEmpty.sort (t2 ^. txMetaOutputs)
+        , t1 ^. txMetaCreationAt == t2 ^. txMetaCreationAt
+        , t1 ^. txMetaIsLocal == t2 ^. txMetaIsLocal
+        , t1 ^. txMetaIsOutgoing == t2 ^. txMetaIsOutgoing
+        ]
+
+
+data InvariantViolation =
+        DuplicatedTransactionWithDifferentHash Core.TxId
+        -- ^ When attempting to insert a new 'MetaTx', the 'Core.TxId'
+        -- identifying this transaction was already present in the storage,
+        -- but when computing the 'Hash' of two 'TxMeta', these values were not
+        -- the same, meaning somebody is trying to re-insert the same 'Tx' in
+        -- the storage with different values (i.e. different inputs/outputs etc)
+        -- and this is effectively an invariant violation.
+      | UndisputableLookupFailed Core.TxId
+        -- ^ When looking up a transaction which the storage claims to be
+        -- already present as a duplicate, such lookup failed. This is an
+        -- invariant violation because a 'TxMeta' storage is append-only,
+        -- therefore the data cannot possibly be evicted, and should be there
+        -- by definition (or we wouldn't get a duplicate collision in the
+        -- first place).
+      deriving Show
+
+-- | A domain-specific collection of things which might go wrong when
+-- storing & retrieving 'TxMeta' from a persistent storage.
+data TxMetaStorageError =
+      InvariantViolated InvariantViolation
+    -- ^ One of the invariant was violated.
+    | StorageFailure SomeException
+    -- ^ The underlying storage failed to fulfill the request.
+    deriving Show
+
+instance Exception TxMetaStorageError
+
+instance Buildable TxMetaStorageError where
+    build storageErr = bprint shown storageErr
+
+instance Arbitrary TxMeta where
+    arbitrary = TxMeta <$> arbitrary
+                       <*> arbitrary
+                       <*> uniqueElements
+                       <*> uniqueElements
+                       <*> arbitrary
+                       <*> arbitrary
+                       <*> arbitrary
+
+-- | Generates 'NonEmpty' collections which do not contain duplicates.
+-- Limit the size to ~10 elements
+uniqueElements :: Gen (NonEmpty (Core.Address, Core.Coin))
+uniqueElements = do
+    (e :| es) <- arbitrary
+    return (e :| (List.take 1 $ List.filter (/= e) (List.nub es)))
```

Comment:
Why not generate a `Set (Core.Address, Core.Coin)` of the desired size and turn it into a `NonEmpty`?